samples/csharp/getting-started/DatabaseIntegration/DatabaseIntegration/Program.cs

```diff
@@ -0,0 +1,102 @@
+using Microsoft.ML;
+using Microsoft.ML.Data;
+using Microsoft.ML.Transforms.Categorical;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+
+namespace DatabaseIntegration
+{
+    public class Program
+    {
+        private static string Download(string baseGitPath, string dataFile)
+        {
+            using (WebClient client = new WebClient())
+            {
+                client.DownloadFile(new Uri($"{baseGitPath}"), dataFile);
+            }
+
+            return dataFile;
+        }
+
+        public static string DownloadAdultDataset()
+            => Download("https://raw.githubusercontent.com/dotnet/machinelearning/244a8c2ac832657af282aa312d568211698790aa/test/data/adult.train", "adult.txt");
+
+        public static List<AdultCensus> GetAdultData()
+        {
+            string adultDataset = DownloadAdultDataset();
+            string dataLine;
+            var dataset = new List<AdultCensus>();
+            using (StreamReader reader = new StreamReader(adultDataset))
+            {
+                // read the header
+                dataLine = reader.ReadLine();
+                while ((dataLine = reader.ReadLine()) != null)
+                {
+                    if (string.IsNullOrEmpty(dataLine))
+                        continue;
+
+                    var data = dataLine.Split(',');
+                    dataset.Add(new AdultCensus()
+                    {
+                        Age = int.Parse(data[0]),
+                        Workclass = data[1],
+                        Education = data[3],
+                        //EducationNum = int.Parse(data[4]),
+                        MaritalStatus = data[5],
+                        Occupation = data[6],
+                        Relationship = data[7],
+                        Race = data[8],
+                        Sex = data[9],
+                        CapitalGain = data[10],
+                        CapitalLoss = data[11],
+                        HoursPerWeek = int.Parse(data[12]),
+                        NativeCountry = data[13],
+                        Label = int.Parse(data[14])
+                    });
+                }
+            }
+
+            return dataset;
+        }
+
+        public static void PopulateDatabase()
+        {
+            var adultData = GetAdultData();
+            using (var db = new AdultCensusContext())
+            {
+                foreach(var row in adultData)
```

Comment:
I still prefer the option of adding things directly to the DbContext and skipping the list, but if you have a list, or an IEnumerable (e.g. a LINQ query that reads the stream and projects AdultCensus instances with a Select(), you can replace this loop with simply db.AdultCensus.AddRange(adultData).

Reply:
Take a look at what I have now and see if this is acceptable as this is much more simplified. So I was not able to put the insert into the Select as the insert (or AddRange) is needed for the whole thing to become "realized". So created the data variable is just an enumerable and wont be evaluated until addrange is called.

---
In reply to: [270255775](https://github.com/dotnet/machinelearning-samples/pull/295#discussion_r270255775) [](ancestors = 270255775)